### PR TITLE
dim statement support

### PR DESCRIPTION
**Pull request by chrisdp**
Having some issues with a few tests @TwitchBronBron if you're willing to give me a had at some point.

### CODE REVIEW

File: src/parser/Parser.ts [hunk 1]
```diff
@@ -1364,6 +1368,38 @@ export class Parser {
         return new ThrowStatement(throwToken, expression);
     }
 
+    private dimStatement() {
+        const dim = this.advance();
+
+        //get the class name
```
Comment (TwitchBronBron) [inline src/parser/Parser.ts:1374]: I think this is the wrong comment. There's nothing to do with class here.

Comment (chrisdp) [reply src/parser/Parser.ts]: Removed

File: src/DiagnosticMessages.ts [hunk 2]
```diff
@@ -605,6 +605,21 @@ export let DiagnosticMessages = {
         message: `Missing exception expression after 'throw' keyword`,
         code: 1118,
         severity: DiagnosticSeverity.Error
+    }),
+    missingLeftBracketAfterDimIdentifier: () => ({
+        message: `Missing left bracket after 'dim' identifier`,
```
Comment (TwitchBronBron) [inline src/DiagnosticMessages.ts:610]: Change function name and message to say "left square bracket" to line up with the way we name our TokenKind (Paren, SquareBracket, CurlyBrace) https://github.com/rokucommunity/brighterscript/blob/master/src/lexer/TokenKind.ts#L5

File: src/DiagnosticMessages.ts [hunk 3]
```diff
@@ -605,6 +605,21 @@ export let DiagnosticMessages = {
...
+        code: 1119,
+        severity: DiagnosticSeverity.Error
+    }),
+    missingRightBracketAfterDimIdentifier: () => ({
+        message: `Missing right bracket after 'dim' identifier`,
```
Comment (TwitchBronBron) [inline src/DiagnosticMessages.ts:615]: Change function name and message to say "left square bracket"

File: src/parser/Parser.ts [hunk 4]
```diff
@@ -1375,6 +1379,44 @@ export class Parser {
...
+                    this.advance();
+                } else {
+                    // will also exit for right square braces
+                    break;
+                }
+            } catch (error) {
+            }
+        } while (expression);
+
+        if (!expressions.length) {
```
Comment (TwitchBronBron) [inline src/parser/Parser.ts:1409]: be explicit here. `if(expressions.length === 0)`

File: src/parser/Statement.ts [hunk 5]
```diff
@@ -591,6 +591,37 @@ export class PrintStatement extends Statement {
...
+    public transpile(state: TranspileState) {
+        const result = [
+            state.sourceNode(this.dimToken, 'dim'),
+            ' '
+        ];
+        return result;
+    }
+
+    public walk(visitor: WalkVisitor, options: WalkOptions) {
+        // if (this.expression && options.walkMode & InternalWalkMode.walkExpressions) {
```
Comment (TwitchBronBron) [inline src/parser/Statement.ts:619]: @elsassph can speak more to this, but I'm pretty sure we need to use the commented out if statement you have, and walk each of the dimensions. So something to the effect of 

```
 if (this.expression && options.walkMode & InternalWalkMode.walkExpressions) {
    for(let i = 0; i < this.dimensions.length; i++) {
        walk(this.dimensions, i, visitor, options, this);
    }
}
```

File: src/parser/Statement.ts [hunk 6]
```diff
@@ -591,6 +591,37 @@ export class PrintStatement extends Statement {
...
+        super();
+        this.range = util.createRangeFromPositions(
+            this.dimToken.range.start,
+            (this.closingSquare ?? this.dimensions[this.dimensions.length - 1] ?? this.openingSquare ?? this.identifier ?? this.dimToken).range.end
+        );
+    }
+    public range: Range;
+
+    public transpile(state: TranspileState) {
+        const result = [
```
Comment (TwitchBronBron) [inline src/parser/Statement.ts:611]: This needs implemented. Something like this:
```typescript
const result = [
    state.sourceNode(this.dimToken, 'dim'),
    ' ',
    state.sourceNode(this.identifier, this.identifier.text),
    state.sourceNode(this.openingSquare, '[')
];
for (var i = 0; i < this.dimensions.length; i++) {
    if (i > 0) {
        result.push(', ');
    }
    result.push(
        ...this.dimensions[i].transpile(state)
    );
}
result.push(state.sourceNode(this.closingSquare, ']'));
```

File: src/parser/Statement.ts [hunk 7]
```diff
@@ -591,6 +591,37 @@ export class PrintStatement extends Statement {
...
+    ) {
+        super();
+        this.range = util.createRangeFromPositions(
+            this.dimToken.range.start,
+            (this.closingSquare ?? this.dimensions[this.dimensions.length - 1] ?? this.openingSquare ?? this.identifier ?? this.dimToken).range.end
+        );
+    }
+    public range: Range;
+
+    public transpile(state: TranspileState) {
```
Comment (TwitchBronBron) [inline src/parser/Statement.ts:610]: Please add some tests proving that this transpiles properly. You can see some transpile test examples [here](https://github.com/rokucommunity/brighterscript/blob/45a31b18b10fecf6ed1046e4e249c5df5a4fcc66/src/files/BrsFile.spec.ts#L1551)

**Review by TwitchBronBron — CHANGES_REQUESTED** [review]
Thanks for all your effort on this. It looks like a good start. Just a few things needed yet to get it across the finish line. Thanks.

**Comment by chrisdp** [comment]
> Thanks for all your effort on this. It looks like a good start. Just a few things needed yet to get it across the finish line. Thanks. 

Thanks. I agree with all the comments. Will address them.

**Pull request merged**